File: Segundo_Curso/Segundo_Cuatri/PDOO/java/src/irrgarten/Labyrinth.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;

/**
 *
 * @author el_dramas
 */
public class Labyrinth {
    static private final int INVALID_POS=-1;
    static private final char BLOCK_CHAR = 'X';
    static private final char EMPTY_BLOCK = '-';
    static private final char MONSTER_CHAR = 'M';
    static private final char COMBAT_CHAR = 'C';
    static private final char EXIT_CHAR = 'E';
    static private final int ROW = 0;
    static private final int COL = 1;
    
    private int nRows;
    private int nCols;
    private int exitRow;
    private int exitCol;
    private Monster mtab[][];
    private char ltab[][];
    private Player ptab[][];
    
    /**
     * Construye un laberinto
     * @param nRows Numero de filas
     * @param nCols Numero de columnas
     * @param exitRow Fila en la que poner la salida
     * @param exitCol Columna en la que poner la salida
     */
    public Labyrinth(int nRows, int nCols, int exitRow, int exitCol){
        this.nRows = nRows;
        this.nCols = nCols;
        this.exitCol = exitCol;
        this.exitRow = exitRow;
        
        mtab = new Monster[nRows][nCols];
        ltab = new char[nRows][nCols];
        ptab = new Player[nRows][nCols];
        
        for (int i=0; i<nRows; ++i){
            for(int j=0; j<nCols; ++j){
                setBlock(i,j,EMPTY_BLOCK);
                setMonsterAt(i,j,null);
                setPlayerAt(i,j,null);
            }
        }
        
        if(this.posOK(exitRow, exitCol))
            setBlock(exitRow, exitCol, EXIT_CHAR);
        else
            setBlock(0,0,EXIT_CHAR);; 
        
    }
    
    /**
     * Reparte los jugadores por el tablero
     * @param players Los jugadores a repartir
     */
    public void spreadPlayers(ArrayList<Player> players){
        for (Player player : players) {
            ArrayList<Integer> newPos = randomEmptyPos();
            //System.out.println(newPos);
            putPlayer2D(INVALID_POS, INVALID_POS, newPos.get(ROW), newPos.get(COL), player);
        }
    }
    
    /**
     * Comprueba si hay un jugador en la casilla de salida
     * @return True si hay un jugador
     */
    public boolean haveAWinner(){
        return getPlayerAt(exitRow, exitCol) != null;
    }
    
    public String toString(){
        String cad = "Labyrinth: \n";
        
        cad += this.convertToString(ltab, nRows, nCols) + "\n\n";
        
        //cad += "Monsters: \n";
        
        //cad += this.convertToString(mtab, nRows, nCols) + "\n\n";
        
        //cad += "Players: \n";
        
        //cad += this.convertToString(ptab, nRows, nCols) + "\n\n";
        
        return cad;
    }
    
    /**
     * Añade un monstruo a la posicion (row, col)
     * @param row fila
     * @param col columna
     * @param monster El mountruo a añadir
     */
    public void addMonster(int row, int col, Monster monster){
        if(this.posOK(row, col)&&(getBlock(row, col)==EMPTY_BLOCK) && !(this.exitPos(row, col))){
            setMonsterAt(row,col,monster);
            setBlock(row,col,MONSTER_CHAR);
            monster.setPos(row, col);
        }     
           
    }
    
    /**
     * Mueve el jugador a una casilla adyacente
     * @param direction Direccion en la que intentar ponerlo
     * @param player El jugador a mover
     * @return Un monstruo si se lo encuentra en la casilla, si no devuelve null
     */
    public Monster putPlayer(Directions direction, Player player){
        int oldRow = player.getRow();
        int oldCol = player.getCol();
        ArrayList<Integer> newPos = dir2Pos(oldRow, oldCol, direction);
        return putPlayer2D(oldRow, oldCol, newPos.get(ROW), newPos.get(COL), player);
    }
    
    /**
     * Añade bloques hasta encontrarse con un obstaculo o llegar al numero deseado
     * @param orientation Si ponerlos en horizontal o vertical
     * @param startRow Fila en la que poner el primer bloque
     * @param startCol Columna en la que poner el primer bloque
     * @param length Numero de bloques a poner
     */
    public void addBlock(Orientation orientation, int startRow, int startCol, int length){
        int incRow=0, incCol=0;
        if(orientation == Orientation.VERTICAL){
            incRow++;
        } else{
            incCol++;
        }
        int row = startRow, col = startCol;
        while( posOK(row,col) && emptyPos(row, col) && length > 0){
            setBlock(row, col, BLOCK_CHAR);
            length--;
            row += incRow;
            col += incCol;
        }

    }
    
    /**
     * Comprueba que movimientos son válidos
     * @param row la fila
     * @param col la columna
     * @return Los movimientos válidos
     */
    public ArrayList<Directions> validMoves(int row, int col){

        ArrayList<Directions> directions = new ArrayList<>(Arrays.asList(Directions.DOWN, Directions.UP, Directions.RIGHT, Directions.LEFT)); //Inicializa la lista con los valores pasados por argumentos
        Iterator<Directions> it = directions.listIterator();  //IMPORTANTE
        
        //Depende del número de direcciones
        while(it.hasNext()){
            ArrayList<Integer> posToCheck = dir2Pos(row, col, it.next());
            if(!canStepOn(posToCheck.get(ROW),posToCheck.get(COL))){
                it.remove();  
            }
        }
        return directions;

    }

    public void turnFuzzy(FuzzyPlayer player){
        setPlayerAt(player.getRow(),player.getCol(), player);
    }
    
    /**
     * Comprueba si la posicion esta dentro del laberinto
     * @param row la fila
     * @param col la columna
     * @return True si esta dentro
     */
    private boolean posOK(int row, int col){
        boolean dentroRow = (0 <= row) && (row<this.nRows);
        boolean dentroCol = (0 <= col) && (col<this.nCols);
        return (dentroRow && dentroCol);
    }
    
    /**
     * Comprueba si una posición esta vacia
     * @param row fila 
     * @param col columna
     * @return True si la posición esta vacia
     */
    private boolean emptyPos(int row, int col){
        return (this.posOK(row,col))&&(getBlock(row, col) == EMPTY_BLOCK);
    }
    
    private boolean monsterPos(int row, int col){
        return (this.posOK(row,col))&&(getBlock(row, col) == MONSTER_CHAR);
    }
    
    private boolean exitPos(int row, int col){
        return (this.posOK(row,col))&&(getBlock(row, col)==EXIT_CHAR);
    }
    
    private boolean combatPos(int row, int col){
        return (this.posOK(row,col))&&(getBlock(row, col) == COMBAT_CHAR);
    }
    
    private boolean canStepOn(int row, int col){
        return (this.posOK(row, col))&& ((this.exitPos(row, col))||(this.emptyPos(row, col)) || (this.monsterPos(row, col)));
    }
    
    /**
     * Actualiza una casilla tras moverse un jugador
     * @param row fila en la que estaba el jugador
     * @param col columna en la que estaba el jugador
     */
    private void updateOldPos(int row, int col){
        if(this.posOK(row, col)){
            if(this.combatPos(row, col))
                setBlock(row, col, MONSTER_CHAR);
            else
                setBlock(row, col, EMPTY_BLOCK);
        }
        
    }
    
    /**
     * Traduce una dirección a una coordenada
     * @param row fila en la que esta el jugador
     * @param col columna en la que esta el jugador
     * @param direction direccion a la que mover
     * @return la nueva posicion
     */
    private ArrayList<Integer> dir2Pos(int row, int col, Directions direction){
        int nextRow=row;
        int nextCol=col;
        switch (direction){
            case LEFT:
                nextCol--;
                break;
            case RIGHT:
                nextCol++;
                break;
            case UP:
                nextRow--;
                break;
            case DOWN:
                nextRow++;
                break;
        }
        
        
        ArrayList<Integer> position = new ArrayList<Integer>();
        position.add(nextRow); //Añade en pos ROW = 0
        position.add(nextCol);  //Añade en pos COL = 1
        return position;
    }
    
    /**
     * Busca una posición vacia en el tablero
     * @return la posición que encuentre primero
     */
    private ArrayList<Integer> randomEmptyPos(){
        Integer row=INVALID_POS,col=INVALID_POS;
        
        while(!(this.posOK(row, col) && this.emptyPos(row, col))){
            row=Dice.randomPos(nRows);
            col=Dice.randomPos(nCols);
        }
        
        ArrayList<Integer> position= new ArrayList<Integer>();
        position.add(ROW,row);
        position.add(COL,col);
        return position;
    }
    
    /**
     * Mueve al jugador y procura dejar el tablero en estado consistente
     * @param oldRow fila antigua
     * @param oldCol columna antigua
     * @param row fila nueva
     * @param col columna nueva
     * @param player El jugador que mover
     * @return Un monstruo si al mover al jugador encuentra a uno, null si no
     */
    private Monster putPlayer2D(int oldRow, int oldCol, int row, int col, Player player){
        Monster output = null;
        if(canStepOn(row, col)){
            if(posOK(oldRow, oldCol)){
                Player p = getPlayerAt(oldRow, oldCol);
                if(p == player){
                    updateOldPos(oldRow, oldCol);
                    setPlayerAt(oldRow, oldCol, null);
                }
            }

            if(monsterPos(row, col)){
                setBlock(row, col, COMBAT_CHAR);
                output = getMonsterAt(row, col);
            } else{
                setBlock(row, col, player.getNumber());
            }

            setPlayerAt(row, col, player);
            player.setPos(row, col);            
        }
        return output;
    }
    
    private String convertToString(char matrix[][], int rows, int cols){
        String cad="";
        
        for (int i=0; i<rows; ++i){
            
            if(i==0){
                for (int j=0; j<3*cols; ++j)
                    cad +="=";
                cad += "\n";
            }
            for (int j=0; j<cols; ++j)
                cad += "|"+matrix[i][j]+"|";
            
            cad += "\n";
            if(i==rows-1){
                for(int j=0; j<3*cols; ++j)
                    cad+="=";
                cad += "\n";
            }
        }
        return cad;
    }

    private void setBlock(int row, int col, char bloque){
        ltab[row][col] = bloque;
    }

    private char getBlock(int row, int col){
        return ltab[row][col];
    }

    private void setMonsterAt(int row, int col, Monster monster){
        mtab[row][col] = monster;
    }

    private Monster getMonsterAt(int row, int col){
        return mtab[row][col];
    }

    
    private void setPlayerAt(int row, int col, Player player){
        ptab[row][col] = player;
    }

    private Player getPlayerAt(int row, int col){
        return ptab[row][col];
    }

}

File: Segundo_Curso/Segundo_Cuatri/PDOO/ruby/Irrgarten/Labyrinth.rb
#encoding: UTF-8

require_relative 'Dice'
require_relative 'Directions'
require_relative 'FuzzyPlayer'
require_relative 'Orientation'

module Irrgarten

    class Labyrinth

        @@INVALID_POS=-1
        @@BLOCK_CHAR='X'
        @@EMPTY_BLOCK='-'
        @@MONSTER_CHAR='M'
        @@COMBAT_CHAR='C'
        @@EXIT_CHAR='E'
        @@ROW=0
        @@COL=1

        # Inicializador
        # nRows->Numero de filas del laberinto
        # nCols->Numero de columnas del laberinto
        # exitRow->Fila de la salida
        # exitCol->Columna de la salida
        def initialize(nRows,nCols,exitRow,exitCol)
            @nRows=nRows
            @nCols=nCols
            @exitRow=exitRow
            @exitCol=exitCol
            @mtab=Array.new(nRows) {(Array.new(nCols))}     #Declaración de un array de arrays de nRows filas y nCols columnas
            @ltab=Array.new(nRows) {(Array.new(nCols))}
            @ptab=Array.new(nRows) {(Array.new(nCols))}     # Se inicializan a nil por defecto

            for i in 0...nRows do
                for j in 0...nCols do
                    @ltab[i][j] = @@EMPTY_BLOCK
                end
            end

            @ltab[exitRow][exitCol] = @@EXIT_CHAR
        end

        # Spawnea los jugadores en el laberinto de forma aleatoria, solo aparecen en bloques vacios
        # Delega en PutPlayer2D
        def spreadPlayers(players)
            for i in 0..players.size-1 do
                pos = randomEmptyPos
                putPlayer2D(@@INVALID_POS,@@INVALID_POS,pos[@@ROW],pos[@@COL],players[i])
            end
        end

        # True -> Hay ganador
        # False -> No hay ganador
        def haveAWinner #bool
            @ptab[@exitRow][@exitCol] != nil
        end

        def to_s
            str="[#{@nRows}x#{@nCols}],(#{@exitRow},#{@exitCol}) \n"
            for i in 0...@nRows do
                for j in 0...@nCols do
                    str += "|" + @ltab[i][j].to_s + "|"
                end
                str +="\n"
            end
            return str
        end

        # Añade un monstruo en una posicion del laberinto, lo añade en el laberinto como una M y en la matriz de monstruos
        # row -> fila del monstruo
        # col -> columna del monstruo
        # monster -> monstruo a añadir
        def addMonster(row,col,monster) #void
            if @ltab[row][col] == @@EMPTY_BLOCK then
                @ltab[row][col] = @@MONSTER_CHAR
                @mtab[row][col] = monster
            end
        end

        # Mueve al jugador a una casilla adyacente
        # direction -> direccion de movimiento
        # player -> jugador a mover
        # return monster -> si hay un monstruo en la posicion de llegada devuelve ese monstruo, si no lo hay devuelve nil
        # post -> La casilla a la que se mueva no tiene por qué ser la que se busca
        def putPlayer(direction, player) #monster
            oldRow = player.row
            oldCol = player.col
            newPos = dir2Pos(oldRow,oldCol,direction)
            monster = putPlayer2D(oldRow,oldCol,newPos[@@ROW],newPos[@@COL],player)
            monster
        end

        # Añade un bloque o conjunto de bloque segun:
        # orientation -> orientacion que se sigue para añadir los bloques
        # startRow -> fila de inicio
        # startCol -> columna de inicio
        # length -> longitud del conjunto de bloques a añadir
        def addBlock(orientation,startRow,startCol, length)#void
            if orientation == Orientation::VERTICAL then
                incRow = 1
                incCol = 0
            else
                incRow = 0
                incCol = 1
            end
            row = startRow
            col = startCol

            while posOK(row,col) && emptyPos(row,col) && length > 0
                @ltab[row][col] = @@BLOCK_CHAR
                length -= 1
                row += incRow
                col += incCol
            end

        end

        # Devuelve un array con las direcciones de movimientos posibles de un jugador
        # row -> fila del jugador
        # col -> columna del jugador
        # output -> Array con las direcciones posibles
        def validMoves(row,col)#Directions[]
            output = Array.new
            if(canStepOn(row+1,col))then
                output.append(Directions::DOWN)
            end

            if(canStepOn(row-1,col))then
                output.append(Directions::UP)
            end

            if(canStepOn(row,col+1))then
                output.append(Directions::RIGHT)
            end

            if(canStepOn(row,col-1))then
                output.append(Directions::LEFT)
            end
            output
        end

        def switchPlayer(player)
            @ptab[player.row][player.col] = player
            @ltab[player.row][player.col] = player.number
        end

	private
        def posOK(row,col)#bool
            (0 <= row && row < @nRows && 0 <= col && col < @nCols)
        end

        def emptyPos(row,col)#bool
            @ltab[row][col] == @@EMPTY_BLOCK
        end

        def monsterPos(row,col)#bool
            @ltab[row][col] == @@MONSTER_CHAR
        end

        def exitPos(row,col)#bool
            @ltab[row][col] == @@EXIT_CHAR
        end

        def combatPos(row,col)#bool
            @ltab[row][col] == @@COMBAT_CHAR
        end

        def canStepOn(row,col)#bool
            posOK(row,col) && (emptyPos(row,col) || monsterPos(row,col) || exitPos(row,col))
        end

        # Actualiza la posicion del tablero despues de combate
        # row -> fila
        # col -> columna
        def updateOldPos(row,col)#void
            if posOK(row, col)
                if @ltab[row][col] == @@COMBAT_CHAR then
                    @ltab[row][col] = @@MONSTER_CHAR
                else
                    @ltab[row][col] = @@EMPTY_BLOCK
                end
            end
        end

        # Traduce una direccion a una coordenada
        # row -> fila
        # col -> columna
        # direction -> direccion de movimiento
        # devuelve un array de dos enteros que representa la posicion
        def dir2Pos(row,col, direction) #int[]
            case direction      #switch
            when Directions::UP
                [row-1, col]
            when Directions::DOWN
                [row+1, col]
            when Directions::LEFT
                [row, col-1]
            when Directions::RIGHT
                [row, col+1]
            end
        end

        # Busca una posicion vacia de forma aleatoria en el tablero
        # Devuelve una posicion en un array de dos posiciones
        def randomEmptyPos #int[]
            begin
                row = Dice.randomPos(@nRows)
                col = Dice.randomPos(@nCols)
            end until emptyPos(row,col)
            [row,col]

        end

        # Mueve el jugador y procura dejar el tablero en estado consistente
        # oldRow -> Fila inicial
        # oldCol -> Columna inicial
        # row -> nueva fila
        # col -> nueva columna
        # player -> nuevo jugador
        # Devuelve un monstruo si lo hay, si no devuelve nil
        def putPlayer2D(oldRow, oldCol, row, col, player) #monster
            output = nil
            if canStepOn(row,col) then
                if posOK(oldRow,oldCol) then
                    p = @ptab[oldRow][oldCol]
                    if p == player then
                        updateOldPos(oldRow,oldCol)
                        @ptab[oldRow][oldCol] = nil
                    end
                end

                monsterPos = monsterPos(row,col)
                if monsterPos then
                    @ltab[row][col] = @@COMBAT_CHAR
                    output = @mtab[row][col]
                else
                    number = player.number
                    @ltab[row][col] = number
                end
                @ptab[row][col] = player
                player.setPos(row,col)
            end
            output
        end

    end#class
end#module
